agent/bootstrap.go

```diff
@@ -0,0 +1,845 @@
+	var repoSSHKeySwitcherRegex = regexp.MustCompile(`-[a-z0-9\-]+$`)
+	host := repoSSHKeySwitcherRegex.ReplaceAllString(url.Host, "")
+
+	// Try and open the existing hostfile in (append_only) mode
+	f, err := os.OpenFile(knownHostPath, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		warningf("Could not open \"%s\" for reading (%s)", knownHostPath, err)
+		return
+	}
+	defer f.Close()
+
+	// Figure out where the ssh tools exist
+	sshToolBinaryPath := ""
+
+	// Grab the generated keys for the repo host
+	keygenOutput, err := b.runCommandSilentlyAndCaptureOutput(filepath.Join(sshToolBinaryPath, "ssh-keygen"), "-f", knownHostPath, "-F", host)
+	if err != nil {
+		warningf("Could not performn `ssh-keygen` (%s)", err)
+		return
+	}
+
+	// If the keygen output doesn't contain the host, we can skip!
+	if !strings.Contains(keygenOutput, host) {
+		// Scan the key and then write it to the known_host file
+		keyscanOutput, err := b.runCommandSilentlyAndCaptureOutput(filepath.Join(sshToolBinaryPath, "ssh-keyscan"), host)
+		if err != nil {
+			warningf("Could not perform `ssh-keyscan` (%s)", err)
+			return
+		}
+
+		if _, err = f.WriteString(keyscanOutput + "\n"); err != nil {
+			warningf("Could not write to \"%s\" (%s)", knownHostPath, err)
+			return
+		}
+
+		commentf("Added \"%s\" to the list of known hosts at \"%s\"", host, knownHostPath)
+	}
+}
+
+// Executes a hook and applyes any environment changes. The tricky thing with
+// hooks is that they can modify the ENV of a bootstrap. And it's impossible to
+// grab the ENV of a child process before it finishes, so we've got an awesome
+// ugly hack to get around this.  We essentially have a bash script that writes
+// the ENV to a file, runs the hook, then writes the ENV back to another file.
+// Once all that has finished, we compare the files, and apply what ever
+// changes to our running env. Cool huh?
+func (b *Bootstrap) executeHook(name string, path string, exitOnError bool) int {
+	hookPath := normalizeScriptFileName(path)
+
+	// Check if the hook exists
+	if fileExists(hookPath) {
+		// Create a temporary file that we'll put the hook runner code in
+		tempHookRunnerFile := createTempFile(normalizeScriptFileName("buildkite-agent-bootstrap-hook-runner"))
+
+		// Ensure the hook script is executable
+		addExecutePermissiontoFile(tempHookRunnerFile.Name())
+
+		// We'll pump the ENV before the hook into this temp file
+		tempEnvBeforeFile := createTempFile("buildkite-agent-bootstrap-hook-env-before")
+		tempEnvBeforeFile.Close()
+
+		// We'll then pump the ENV _after_ the hook into this temp file
+		tempEnvAfterFile := createTempFile("buildkite-agent-bootstrap-hook-env-after")
+		tempEnvAfterFile.Close()
+
+		absolutePathToHook, err := filepath.Abs(hookPath)
+		if err != nil {
+			fatalf("Failed to find absolute path to \"%s\" (%s)", hookPath, err)
+		}
+
+		// Create the hook runner code
+		var hookScript string
+		if runtime.GOOS == "windows" {
+			hookScript = "@echo off\n" +
```

Comment:
You should do this using a golang template as illustrated below as it will save the mess of escaping.

see https://github.com/wolfeidau/aws-ek-setup/blob/master/cmd/aws-ek-setup/command_create.go#L27-L39 and executed https://github.com/wolfeidau/aws-ek-setup/blob/master/cmd/aws-ek-setup/command_create.go#L75-L92